src/memory_coloring.cpp

```diff
@@ -0,0 +1,381 @@
+                it = std::max_element(segments.begin(), segments.end(), [&](segment x, segment y) {
+                    return x.second < y.second;
+                });
+            if(it != segments.end())
+                start = it->second;
+        }
+        auto s = segment{start, start + n};
+        assert(not overlaps(segments, s));
+        segments.insert(s);
+        return s;
+    }
+
+    // Build the allocation_color class from the conflict_table
+    static allocation_segment build(const instruction_set_map& conflict_table,
+                                    std::size_t alignment)
+    {
+        allocation_segment as{};
+        std::vector<instruction_ref> conflict_queue;
+        // Add all allocations to the conflict_queue
+        std::transform(conflict_table.begin(),
+                       conflict_table.end(),
+                       std::back_inserter(conflict_queue),
+                       [](auto&& pp) { return pp.first; });
+
+        // Sort the conflict queue so we process the allocation with the least
+        // number of adjacent allocations first
+        std::sort(conflict_queue.begin(), conflict_queue.end(), [&](auto x, auto y) {
+            return std::make_tuple(conflict_table.at(x).size(), x->get_shape().bytes()) <
+                   std::make_tuple(conflict_table.at(y).size(), y->get_shape().bytes());
+        });
+        // Process the conflict_queue, we refer to the current allocation as
+        // the parent and the adjacent allocations as children
+        for(auto parent : conflict_queue)
+        {
+            // Sort children by size
+            std::vector<instruction_ref> children(conflict_table.at(parent).begin(),
+                                                  conflict_table.at(parent).end());
+            std::sort(children.begin(), children.end(), [](auto x, auto y) {
+                return x->get_shape().bytes() < y->get_shape().bytes();
+            });
+            assert(not contains(children, parent));
+            // This set is to track the segments already processed
+            std::set<segment> segments;
+            // Add all segments for the children to the segments already processed
+            transform_if(
+                children.begin(),
+                children.end(),
+                std::inserter(segments, segments.begin()),
+                [&](auto child) { return as.get_segment(child); },
+                [&](auto child) { return *as.get_segment(child); });
+
+            // Get the segment for the parent
+            const auto* parent_segment = as.get_segment(parent);
+            // Add segment for the parent if there is none or segment overlaps with the children
+            if(parent_segment == nullptr or overlaps(segments, *parent_segment))
+                as.add_segment(parent, next_segment(segments, parent, alignment));
+        }
+        // Reduce the number of segments
+        for(std::size_t n = 0; n < 3; n++)
```

Comment:
count of 3, is it arbitrary ?

Reply:
Yea, we loop three time, it seems to improve memory usage.

Reply:
okay, just by looking at the code, it seems loop is guaranteed to converge after a certain iterations (i could be wrong), if that is the case then,  we can do a while loop with some exit condition.

Reply:
It wont converge. Its NP-Hard so we wont know if its correct unless we try all options.